File: backend/app/schemas/billing.py
"""Billing schemas."""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class PlanResponse(BaseModel):
    """Plan response schema."""

    id: UUID
    name: str
    display_name: str
    description: str | None

    # Pricing (in kopeks)
    price_monthly: int
    price_yearly: int

    # Limits
    max_cron_tasks: int
    max_delayed_tasks_per_month: int
    max_workspaces: int
    max_execution_history_days: int
    min_cron_interval_minutes: int

    # Features
    telegram_notifications: bool
    email_notifications: bool
    webhook_callbacks: bool
    custom_headers: bool
    retry_on_failure: bool

    # Visibility
    is_active: bool
    is_public: bool
    sort_order: int

    class Config:
        from_attributes = True


class SubscriptionResponse(BaseModel):
    """Subscription response schema."""

    id: UUID
    user_id: UUID
    plan_id: UUID
    status: str
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool
    cancelled_at: datetime | None

    # Include plan details
    plan: PlanResponse | None = None

    class Config:
        from_attributes = True


class CreatePaymentRequest(BaseModel):
    """Request to create a payment."""

    plan_id: UUID
    billing_period: str = Field(default="monthly", pattern="^(monthly|yearly)$")
    return_url: str | None = None


class PaymentResponse(BaseModel):
    """Payment response schema."""

    id: UUID
    user_id: UUID
    amount: int
    currency: str
    status: str
    description: str | None
    yookassa_confirmation_url: str | None
    paid_at: datetime | None
    created_at: datetime

    class Config:
        from_attributes = True


class CancelSubscriptionRequest(BaseModel):
    """Request to cancel subscription."""

    immediately: bool = False


class WebhookPayload(BaseModel):
    """YooKassa webhook payload."""

    type: str
    event: str
    object: dict

File: frontend/src/api/billing.ts
import { apiClient } from './client'

export interface Plan {
  id: string
  name: string
  display_name: string
  description: string | null
  price_monthly: number
  price_yearly: number
  max_cron_tasks: number
  max_delayed_tasks_per_month: number
  max_workspaces: number
  max_execution_history_days: number
  min_cron_interval_minutes: number
  telegram_notifications: boolean
  email_notifications: boolean
  webhook_callbacks: boolean
  custom_headers: boolean
  retry_on_failure: boolean
}

export interface Subscription {
  id: string
  workspace_id: string
  plan_id: string
  status: 'active' | 'past_due' | 'cancelled' | 'expired'
  current_period_start: string
  current_period_end: string
  cancel_at_period_end: boolean
  cancelled_at: string | null
  plan?: Plan
}

export interface Payment {
  id: string
  workspace_id: string
  amount: number
  currency: string
  status: 'pending' | 'waiting_for_capture' | 'succeeded' | 'cancelled' | 'refunded'
  description: string | null
  yookassa_confirmation_url: string | null
  paid_at: string | null
  created_at: string
}

export interface PricePreview {
  plan_price: number
  proration_credit: number
  final_amount: number
  remaining_days: number
  currency: string
}

export async function getPlans(): Promise<Plan[]> {
  const response = await apiClient.get('/billing/plans')
  return response.data
}

export async function getSubscription(): Promise<Subscription | null> {
  const response = await apiClient.get('/billing/subscription')
  return response.data
}

export async function createPayment(
  planId: string,
  billingPeriod: 'monthly' | 'yearly' = 'monthly',
  returnUrl?: string
): Promise<Payment> {
  const response = await apiClient.post('/billing/subscribe', {
    plan_id: planId,
    billing_period: billingPeriod,
    return_url: returnUrl,
  })
  return response.data
}

export async function cancelSubscription(
  immediately: boolean = false
): Promise<void> {
  await apiClient.post('/billing/subscription/cancel', {
    immediately,
  })
}

export async function getPaymentHistory(
  limit: number = 20,
  offset: number = 0
): Promise<Payment[]> {
  const response = await apiClient.get('/billing/payments', {
    params: { limit, offset },
  })
  return response.data
}

export async function previewPrice(
  planId: string,
  billingPeriod: 'monthly' | 'yearly' = 'monthly'
): Promise<PricePreview> {
  const response = await apiClient.post('/billing/preview-price', {
    plan_id: planId,
    billing_period: billingPeriod,
  })
  return response.data
}
